include/arch/riscv/arch/32/mode/machine.h

```diff
@@ -8,4 +8,42 @@
 
 /* Place holder for 32-bit machine header */
 
+#include <util.h>
 #include <arch/model/smp.h>
+#include <stdint.h>
+
+static inline uint64_t riscv_read_time(void)
+{
+    register word_t nH_prev, nH, nL;
+    asm volatile(
+        "rdtimeh %0\n"
+        "rdtime  %1\n"
+        "rdtimeh %2\n"
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH)
+        : /* no input */
+        : "memory");
+    if (nH_prev != nH) {
+        /* Ensure that the time is correct if there is a rollover in the
+         * high bits between reading the low and high bits. */
+        asm volatile("rdtime  %0\n" : "=r"(nL) :: "memory");
+    }
+    return ((uint64_t)nH << 32) | nL;
+}
+
+static inline uint64_t riscv_read_cycle(void)
+{
+    register word_t nH_prev, nH, nL;
+    asm volatile(
+        "rdcycleh %0\n"
+        "rdcycle  %1\n"
+        "rdcycleh %2\n"
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH)
+        : /* no input */
+        : "memory");
+    if (nH_prev != nH) {
+        /* Ensure that the time is correct if there is a rollover in the
+         * high bits between reading the low and high bits. */
+        asm volatile("rdtime  %0\n" : "=r"(nL) :: "memory");
```

Comment:
This should be `rdcycle` instead of `rdtime`.

Reply:
thanks, fixed